lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -4400,6 +4400,15 @@ void swift::diagnoseConformanceFailure(Type T,
 
   diags.diagnose(ComplainLoc, diag::type_does_not_conform,
                  T, Proto->getDeclaredType());
+
+  // Special case: Automatic synthesis of Comparable only works with enums for
+  // now. If we are trying to conform to Comparable and failing, emit a note to
+  // let the user know about this.
+  if (!T->getEnumOrBoundGenericEnum() &&
```

Comment:
I have added the check here because we simply do not attempt any synthesis if `derivesProtocolRequirement` returns false. I thought about putting the check there but I didn't want the diagnostic to be a side-effect of performing a check (and `derivesProtocolRequirement` doesn't do any diagnostics anyway and is only to check if it's possible to synthesise the requirement or not).

Let me know if you think this should go elsewhere :)

Reply:
Is there a place where compiler detect that conformance requirements need to be derived automatically? That might be a place to diagnose this...

Reply:
Hmm, I’ve looked around a bit but I couldn’t find a suitable place other than this, because the code for `Comparable` synthesis doesn’t even get invoked in situations like this. 

My understanding is that the compiler resolves witnesses using three strategies - it attempts to resolve it using name lookup, if we fail to find a witness, then it tries to derive it and if we still fail to find it then it tries to find a default implementation.

In the second step (derivation), the compiler first checks if it’s possible to derive the witness (by calling `derivesProtocolRequirement`). If it’s not possible to derive it, it just bails out. 

Now, we _could_ emit the note in `DeriveConformanceComparable.cpp` but to do that would would have to move the “can I derive the `Comparable` requirement?” check from `derivesProtocolRequirement` into `deriveComparable` itself, but I am not sure if that’s the right thing to do. What do you think?

Reply:
What about `DerivedConformance::tryDiagnoseFailedDerivation`? It seems like `resolveWitnessViaLookup` has all the necessary information to determine whether it's even possible to attempt to derive requirements.

Reply:
You're right, looks like I didn't notice that function at all 🤦 I have moved the check there now!